hyperloglog/src/main/java/com/hazelcast/jet/hyperloglog/ProbabilisticAggregations.java

```diff
@@ -0,0 +1,97 @@
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.DataSerializable;
+import com.hazelcast.util.HashUtil;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectOutputStream;
+import java.util.function.Function;
+
+public final class ProbabilisticAggregations {
+    private static final SupplierEx<Function<Object, Long>> DEFAULT_FUNCTION_FN_SUPPLIER = HashFnSuppliers.SERIALIZE_AND_HASH;
+
+    private ProbabilisticAggregations() {
+
+    }
+
+    public static <T> AggregateOperation1<T, ?, Long> hyperLogLog() {
+        return hyperLogLog(DEFAULT_FUNCTION_FN_SUPPLIER);
+    }
+
+    public static <T> AggregateOperation1<T, ?, Long> hyperLogLog(SupplierEx<Function<Object, Long>> hashingFnSupplier) {
+        return AggregateOperation
+                .withCreate(() -> new HyperLogLogAcc<T>(hashingFnSupplier)
+                )
+                .<T>andAccumulate(HyperLogLogAcc::add)
+                .andCombine(HyperLogLogAcc::merge)
+                .andExportFinish(HyperLogLogAcc::estimate);
+    }
+
+    public enum HashFnSuppliers implements SupplierEx<Function<Object, Long>> {
+        SERIALIZE_AND_HASH {
+            @Override
+            public FunctionEx<Object, Long> getEx() throws Exception {
+                ByteArrayOutputStream baos = new ByteArrayOutputStream();
+                ObjectOutputStream oos = new ObjectOutputStream(baos);
+                return (item) -> serializeAndHash(item, baos, oos);
+            }
+        }
+    }
+
+    private static class HyperLogLogAcc<T> implements DataSerializable {
+        private HyperLogLogImpl hyperLogLog = new HyperLogLogImpl();
+        private SupplierEx<Function<Object, Long>> hashingFnSupplier;
+        private transient Function<Object, Long> hashingFn;
+
+        public HyperLogLogAcc(SupplierEx<Function<Object, Long>> hashingFnSupplier) {
+            this.hashingFnSupplier = hashingFnSupplier;
+            this.hashingFn = hashingFnSupplier.get();
+        }
+
+        @Override
+        public void writeData(ObjectDataOutput out) throws IOException {
+            out.writeObject(hyperLogLog);
+            out.writeObject(hashingFnSupplier);
+        }
+
+        @Override
+        public void readData(ObjectDataInput in) throws IOException {
+            hyperLogLog = in.readObject();
+            hashingFnSupplier = in.readObject();
+            hashingFn = hashingFnSupplier.get();
+        }
+
+        void add(T item) {
+            long hash = hashingFn.apply(item);
+            hyperLogLog.add(hash);
+        }
+
+        void merge(HyperLogLogAcc<T> other) {
+            hyperLogLog.merge(other.hyperLogLog);
+        }
+
+        long estimate() {
+            return hyperLogLog.estimate();
+        }
+    }
+
+
+    private static <T> long serializeAndHash(T item, ByteArrayOutputStream baos, ObjectOutputStream oos) throws IOException{
```

Comment:
there are multiple issues with this:
1. performance
2. it will fail for domain objects which implement Hazelcast serialization, but do not implement `j.i.Serializable`

I would need some sort of access to Hazelcast serialization service.

Reply:
I think you can only accept serialized items here and provide a context factory for serialization

Reply:
done